ttnn/cpp/ttnn/operations/ccl/barrier/device/barrier_op.cpp

```diff
@@ -0,0 +1,98 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include "ttnn/operations/ccl/barrier/device/barrier_op.hpp"
+#include "tt_metal/host_api.hpp"
+
+#include <cstdint>
+
+namespace ttnn {
+
+void Barrier::validate(const std::vector<Tensor>& input_tensors) const {
+    //Validate the input tensor
+    TT_FATAL(this->topology == ccl::Topology::Ring, "We currently only support Ring topologies on this OP");
+}
+
+std::vector<SimpleShape> Barrier::compute_output_shapes(const std::vector<Tensor>& input_tensors) const {
+    //For a Barrier the output shape should match the input
+    SimpleShape shape = input_tensors[0].get_logical_shape();
+    return std::vector<SimpleShape>(input_tensors.size(), shape);
+}
+
+std::vector<Tensor> Barrier::create_output_tensors(const std::vector<Tensor>& input_tensors) const {
+    //Tensor is unmodified, return what was passed in
+    return input_tensors;
+}
```

Comment:
Just an FYI, I think we want to move `create_output_tensors` to be fully manual mode for cases like this. This means you shouldn't have to even implement `compute_output_shapes`.